Update for Gnome 45

jtaala opened the pull request:
Fixes #576 

Gnome 45 will require every gnome extension to be updated in order to work in 45. That is, no previous extensions will work on 45 due to moving to ECMAScript modules (meaning that the previous gnome extensions imports won't work).

This PR is for implementing full Gnome 45 support in PaperWM.

Update: It is done - long live Gnome 45!

jtaala committed:
Ported App.js to ESM.

jtaala committed:
WIP: porting `extensions.js`.

jtaala committed:
Merge branch 'release' into gnome-45

jtaala committed:
Ported `extensions.js` to ESM.  Using centralised imports file (which
exports all modules).

jtaala committed:
Ported app.js.

jtaala committed:
Ported gestures.js.  Allow shadowing in eslint.

jtaala committed:
Ported grab.js.

jtaala committed:
Ported settings.js.

jtaala committed:
Ported keybindings.js to ESM imports.

jtaala committed:
Export functions in lib.js.

jtaala committed:
Ported liveAltTab to ESM imports.

jtaala committed:
Ported minimap.js to ESM.

jtaala committed:
Ported Utils.js to ESM.

jtaala committed:
Ported navigator.js.

jtaala committed:
Ported scratch.js.

jtaala committed:
Ported patches.js.

jtaala committed:
Ported workspace.js.

jtaala committed:
Ported topbar.js.

jtaala committed:
Ported stackoverlay.js

jtaala committed:
First attempt at porting tiling.js to ESM.

jtaala committed:
Updated shell version to "45".

jtaala committed:
Lots of import related fixes.

jtaala committed:
Migrated all debug calls to `console.debug`.

jtaala committed:
Removed old/unused debug functions.

jtaala committed:
Fixes for exports and also now that ESM, only parent modules can change
it's exported values (other modules can't).

jtaala committed:
FIXES: exports, and replace Meta.Rectangle.  Also removed several
dynamic function ref wrappers.

jtaala committed:
Removed deprecated `global.screen` (pre gnome 40).

jtaala committed:
Removed `dynamic_funtion_ref` approach and using ESM module references
directly.

jtaala committed:
Improved module dependency ordering for `disable` to avoid actioning a
`workspaces-only-on-primary` call after PaperWM disable.

jtaala committed:
WIP: prefs.js

jtaala committed:
WIP: keybindings_pane, passing keyboard settings schema.

jtaala committed:
Pres.js window improvements.

jtaala committed:
Large pref.js clean and refactored functions into class.

jtaala committed:
Migrated workspace.js to WorkspaceSettings class.

jtaala committed:
Classifying AccleratorParse.

jtaala commented:
It is done - long live Gnome 45!

jtaala committed:
Added note re `user.js` not working Gnome 45.  Now supplying a
`user.css` example in user config directory.

jtaala committed:
Updated ego generation script to include user.css example file.

jtaala committed:
Updated default vertical margins.

jtaala committed:
Update README.md with gnome-45 related stuff.

jtaala committed:
Merge branch 'release' into gnome-45

jtaala commented:
Hey all,

Gnome 45 compatibility is now done.  A Gnome 45 extension was accepted (passed with no issues raised!) by  [extensions.gnome.org](https://extensions.gnome.org/extension/6099/paperwm/) and is now live on ego.

This is a large PR with many things changed.  Main thing to look at is the changes in [`README.md`](https://github.com/paperwm/PaperWM/pull/605/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5) as this gives a heads-up up re some gnome 45 changes (`user.js` not now available in Gnome 45... will look into that at some other time).  Also states re `develop` and `release` branches (will be) on Gnome 45.

Planning on merging this into `develop` and `release` in a few days.  Please have a look and see if other things we should say in `README.md`.

Thanks all.

paradajz commented:
GNOME 45 includes new activity button which has (for me at least) a very useful feature - scrolling while the mouse pointer is located on that button switches the workspaces. Super useful as I was for years now using [Scroll Panel](https://extensions.gnome.org/extension/4257/scroll-panel/) extension. So this is great! However, PaperWM disables that button and replaces it with workspace names. The functionality in PaperWM also enables the scrolling on that, but it's not immediate like the new built-in functionality and it also switches workspaces in most-used order which I don't like (and it's not configurable). So... any way of using built-in button from GNOME?

Lythenas commented on acceleratorparse.js:
```diff
@@ -1,2375 +1,2382 @@
-        ok = false;
-        key = 0;
-        mask = 0;
+export class AcceleratorParse {
```
Is there a point to having this as a class? You are still basically only calling `accelerator_parse` but now you have to pass around an object.

jtaala replied:
Thanks @Lythenas, yes, there is a specific reason for this.  So, it was originally a module - issue with top-level fields in modules is that the populate/pollute global scope in Gnome shell.  Which is okay, but they must be explicitly cleaned on on `disable`.

However, `pref.js` is not running on the gnome-shell process (runs in own/separate process)... and has no `disable` - but it still can populate global Gnome scope.  So, we were using (in 44) a hack to clean this up in `prefs.js` (essentially detected `close-request` events).  Gnome team asked us to change this, especially for Gnome 45.

Hence, it's now an object (inside a `prefs.js` class) that doesn't pollute global scope - and hence doesn't require explicit cleaning up.

Note too, that there's only one `accelerator_parse` object in PaperWM (gnome-shell process) and another one when opening up `prefs` (note there's no comms between PaperWM extension and prefs - although we can share modules or detected changes via `gsettings`).

Anyways, this was recommended and requested to change to object by Gnome team:
![image](https://github.com/paperwm/PaperWM/assets/30424662/192789e1-2b18-4068-a226-c5353697e60c)

Lythenas reviewed (commented):
This looks good to me but I haven't tried it out yet.

I moved the content of `notes.org` to a wiki page just to preserve it: https://github.com/paperwm/PaperWM/wiki/Legacy-notes

jtaala commented:
> GNOME 45 includes new activity button which has (for me at least) a very useful feature - scrolling while the mouse pointer is located on that button switches the workspaces. Super useful as I was for years now using [Scroll Panel](https://extensions.gnome.org/extension/4257/scroll-panel/) extension. So this is great! However, PaperWM disables that button and replaces it with workspace names. The functionality in PaperWM also enables the scrolling on that, but it's not immediate like the new built-in functionality and it also switches workspaces in most-used order which I don't like (and it's not configurable). So... any way of using built-in button from GNOME?

Yes, this looks like pretty cool.  My focus for this branch was just to get it working Gnome 45 so haven't explicitly looked at the new activity button.

Having said that - you can re-enable in `gnome-45` branch it by commenting out this line:
https://github.com/paperwm/PaperWM/blob/8988345faf21f590d7990252befb971d4c6b1d66/topbar.js#L49C6-L49C6

I'll look at adding a preference for it (maybe even with default on).

See also this related issue #604.

jtaala commented:
> GNOME 45 includes new activity button which has (for me at least) a very useful feature - scrolling while the mouse pointer is located on that button switches the workspaces. Super useful as I was for years now using [Scroll Panel](https://extensions.gnome.org/extension/4257/scroll-panel/) extension. So this is great! However, PaperWM disables that button and replaces it with workspace names. The functionality in PaperWM also enables the scrolling on that, but it's not immediate like the new built-in functionality and it also switches workspaces in most-used order which I don't like (and it's not configurable). So... any way of using built-in button from GNOME?

Thanks for raising this @paradajz - looking at this, found an issue given that element has changed (e.g. setting `use-workspace-name` dconf setting to `false` breaks PaperWM in gnome 45 since the new activity button has changed and PaperWM expected some things that are no longer there).

Will push up a fix for this corner case and look at adding an option.

jtaala committed:
FIX: for `use-workspace-name` issue (that can cause PaperWM to not
start).`

jtaala committed:
FIX: if hide workspace indicator `show-workspace-indicator` is false ==>
show Gnome activities instead.

jtaala committed:
Remove `use-workspace-name` deprecated preference.

jtaala committed:
Workspace indicator mouse scrolling now sequence.

jtaala committed:
FIX: for FocusIcon on spaces.

jtaala commented:
> GNOME 45 includes new activity button

Hey @paradajz - please pull the latest on this branch and from a terminal run:

```
dconf write /org/gnome/shell/extensions/paperwm/show-workspace-indicator false
```

Which will disable the PaperWM workspace indicator and use the new Gnome activities element instead.

Let me know how it goes.

jtaala commented:
P.S. I've also changed the scrolling behaviour on PaperWM workspace element to sequence (not stack) - this is to align with Gnome Activities behaviour here (and it makes more sense).

#604.

paradajz commented:
Great. It works. However now I've noticed two things:

1) My shortcuts for moving windows to the next/previous workspace don't work anymore - I am not sure if this PR changed something or did G45 override something. The shortcuts are Ctrl+Shift+Left for moving to the left workspace and Ctrl+Shift+Right to the right. When I press this combination what I get instead is switching between workspaces in order. This combination isn't used anywhere else. I've also tried to change the shortcuts to use PageUp/Down instead of left/right arrows and I get the same issue. Two meta issues here:
a) Since GNOME doesn't have vertical workspaces anymore, maybe the better wording would be to move to workspaces left/right instead of up/down in the preferences
b) The workspace switching animation also uses vertical animation instead of horizontal

2) Every single option in the keybindings section is trimmed:

![image](https://github.com/paperwm/PaperWM/assets/2544094/8ece0ca2-1cc4-4a7e-8a46-35ea0415e7d6)

paradajz commented:
One other thing: With PaperWM enabled and new activity button enabled, scrolling the workspaces doesn't use the GNOME animation. When PaperWM is disabled, animation is present.

jtaala committed:
FIX: moving windows in workspace sequence (regression).

jtaala commented:
> 1. My shortcuts for moving windows to the next/previous workspace don't work anymore

Well, let's looks at this one first.  Can you please pull the latest and test.  Was a regression from an extra `=` that was added (harder equality).

jtaala commented:
> 2\. Every single option in the keybindings section is trimmed

Hmmm - I'm not seeing this, and you're close button is on the left(?) which is different than standard gnome stuff.  Are you running another extension to change these elements / themes?

This is what I'm seeing:

![image](https://github.com/paperwm/PaperWM/assets/30424662/7272e5fa-7d41-4984-9c72-44feb0ace7d7)

jtaala commented:
> One other thing: With PaperWM enabled and new activity button enabled, scrolling the workspaces doesn't use the GNOME animation. When PaperWM is disabled, animation is present.

I was wondering about this, I thought it strange that there was no animation - will have a look at this tomorrow.

paradajz commented:
> > 1. My shortcuts for moving windows to the next/previous workspace don't work anymore
> 
> Well, let's looks at this one first. Can you please pull the latest and test. Was a regression from an extra `=` that was added (harder equality).

Works now! Thanks!

> Hmmm - I'm not seeing this, and you're close button is on the left(?) which is different than standard gnome stuff. Are you running another extension to change these elements / themes?

No, it's just an option in GNOME Tweak Tool:

![image](https://github.com/paperwm/PaperWM/assets/2544094/3c697ecc-99a7-4a61-8d95-f110c3932b3c)

However I do run with scaling factor set to 0.9 for years now on this monitor (looks better to me). Tried to set it to 1 and yeah, now the options aren't trimmed anymore. Not sure if this is something for you to look at then.

![image](https://github.com/paperwm/PaperWM/assets/2544094/5cf84a16-6394-4587-8f82-8585727cfafd)

![image](https://github.com/paperwm/PaperWM/assets/2544094/72bd424f-b9b0-45b1-8858-a431032724ac)

jtaala committed:
FIX #608.

jtaala committed:
Merge branch 'fix-608-overview-minimize-on-close-app-gnome-crash' into gnome-45

jtaala committed:
Merge branch 'fix-608-overview-minimize-on-close-app-gnome-crash' into gnome-45

jtaala committed:
Merge branch 'fix-liveAltTab-size-scales-with-minimap' into gnome-45

jtaala commented:
@Lythenas, I've still got to merge this at some point.

What do you recon?  when we merge this, should be switch back to `develop` branch as default?  Given if users what a more `stable` release they could just use the EGO version?

I think it's ready - although I won't be on Gnome 45 until Arch switches to it (Arch usually waits a bit for Gnome releases).  Still using 44 for my daily driver and dev'ing Gnome 45 in a libvirt virtual machine.

Lythenas commented:
I think merging this into develop is probably fine. I think the message in the readme is clear enough.

For people that are just going to pull on gnome 44 and less this will break the extension. But not sure how we could prevent that.

As for testing I will also switch once Gnome 45 is available on Arch. Before that I can't really test multi monitor support. Everything else seems to work fine in a VM as far as I can tell. But I didn't test that thoroughly.

jtaala committed:
Merge branch 'feature-add-other-preview-scaling-prefs' into gnome-45

jtaala committed:
FIXES to minimap sizing and label placement.

jtaala committed:
FIXES: removed "opacity" actor hack for opening animations.  Essentially
removed opening "scale" animation which results in a cleaner/snappier
opening of apps (while avoiding the clipping issue).

jtaala committed:
Removed opening animation easer.

jtaala committed:
Merge branch 'fix-opening-animation-glitches' into gnome-45

jtaala commented:
Okay, I'm about ready to merge this in and make it default for `develop` and `release` branches.

Note, I'm also now on gnome 45 for Arch (using [FCGU gnome arch repo](https://codeberg.org/fabiscafe/fcgu)).

jtaala committed:
Merge branch 'fix-minimap-sizing' into gnome-45

jtaala committed:
Added option to use Gnome workspace indicator pill.

jtaala committed:
Moved prefs "Settings" box to the top of general tab (most used
settings).
